java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -39,144 +40,145 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public class FlowComputation {
 
   private final Predicate<Constraint> addToFlow;
   private final Predicate<Constraint> terminateTraversal;
-  private final List<JavaFileScannerContext.Location> flow = new ArrayList<>();
-  private final SymbolicValue symbolicValue;
-  private final Set<ExplodedGraph.Node> visited = new HashSet<>();
-
+  private final Set<SymbolicValue> symbolicValues;
+
+  /**
+   *
+   * @param symbolicValue - can be null only in case computation is requested on SV_0 in {@link org.sonar.java.se.checks.NullDereferenceCheck},
+   *                      this is to avoid highlighting all {@code null} literals in the code
+   * @param addToFlow - predicate used to filter constraints which should be added to the flow
+   * @param terminateTraversal - predicate used to terminate flow when specific constraint is detected on the path
+   */
   private FlowComputation(@Nullable SymbolicValue symbolicValue, Predicate<Constraint> addToFlow, Predicate<Constraint> terminateTraversal) {
     this.addToFlow = addToFlow;
     this.terminateTraversal = terminateTraversal;
-    this.symbolicValue = symbolicValue;
+    this.symbolicValues = computedFrom(symbolicValue).collect(Collectors.toSet());
+  }
+
+  private static Stream<SymbolicValue> computedFrom(@Nullable SymbolicValue symbolicValue) {
+    if (symbolicValue == null) {
+      return Stream.empty();
+    }
+    return Stream.concat(Stream.of(symbolicValue), symbolicValue.computedFrom().stream().flatMap(FlowComputation::computedFrom));
   }
 
-  public static List<JavaFileScannerContext.Location> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal) {
+  public static Set<List<JavaFileScannerContext.Location>> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal) {
     return flow(currentNode, currentVal, constraint -> true);
   }
 
-  public static List<JavaFileScannerContext.Location> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow) {
+  public static Set<List<JavaFileScannerContext.Location>> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow) {
     return flow(currentNode, currentVal, addToFlow, c -> false);
   }
 
-  public static List<JavaFileScannerContext.Location> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow,
+  public static Set<List<JavaFileScannerContext.Location>> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow,
     Predicate<Constraint> terminateTraversal) {
     FlowComputation flowComputation = new FlowComputation(currentVal, addToFlow, terminateTraversal);
-
+    // FIXME SONARJAVA-2049 getLastEvaluated doesn't work with relational SV
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)